File: web/lib/clipDelivery.ts
/**
 * Clip delivery API client.
 *
 * Secure endpoints for clip playback, download, and sharing.
 * URLs are short-lived and should be used immediately.
 */

import { toast } from "sonner";

import { apiFetch } from "@/lib/apiClient";

// ============================================================================
// Types
// ============================================================================

/** Access level for shared clips */
export type ShareAccessLevel = "none" | "view_playback" | "download";

/** Summary of clip metadata */
export interface ClipSummary {
  clip_id: string;
  filename: string;
  title: string;
  duration_seconds: number;
  file_size_bytes: number;
}

/** Delivery URL response */
export interface DeliveryUrl {
  url: string;
  expires_at: string;
  expires_in_secs: number;
  content_type: string;
}

/** Playback/download URL response */
export interface PlaybackUrlResponse extends DeliveryUrl {
  clip: ClipSummary;
}

/** Request body for download URL */
export interface DownloadUrlRequest {
  filename?: string;
}

/** Request body for creating a share */
export interface CreateShareRequest {
  access_level?: ShareAccessLevel;
  expires_in_hours?: number;
  watermark_enabled?: boolean;
}

/** Share creation response */
export interface ShareResponse {
  share_url: string;
  share_slug: string;
  access_level: ShareAccessLevel;
  /** When the share expires (ISO 8601). May be undefined if field is omitted from JSON. */
  expires_at?: string | null;
  watermark_enabled: boolean;
  created_at: string;
}

// ============================================================================
// Playback URL
// ============================================================================

/**
 * Get a short-lived playback URL for a clip.
 *
 * Use this URL in a `<video>` element or window.open() for "Play in new tab".
 * URLs expire in ~15 minutes by default.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 * @returns Playback URL response with clip metadata
 *
 * @example
 * ```ts
 * const { url, clip } = await getPlaybackUrl("clip-123", authToken);
 * videoElement.src = url;
 * ```
 */
export function getPlaybackUrl(
  clipId: string,
  token: string
): Promise<PlaybackUrlResponse> {
  return apiFetch<PlaybackUrlResponse>(
    `/api/clips/${encodeURIComponent(clipId)}/play-url`,
    {
      method: "POST",
      token,
    }
  );
}

// ============================================================================
// Download URL
// ============================================================================

/**
 * Get a short-lived download URL for a clip.
 *
 * URLs include Content-Disposition headers to prompt browser download.
 * URLs expire in ~5 minutes by default.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 * @param options - Optional download options (custom filename)
 * @returns Download URL response with clip metadata
 *
 * @example
 * ```ts
 * const { url } = await getDownloadUrl("clip-123", authToken, {
 *   filename: "my-awesome-clip.mp4"
 * });
 * // Trigger download
 * const a = document.createElement('a');
 * a.href = url;
 * a.download = "my-awesome-clip.mp4";
 * a.click();
 * ```
 */
export function getDownloadUrl(
  clipId: string,
  token: string,
  options?: DownloadUrlRequest
): Promise<PlaybackUrlResponse> {
  return apiFetch<PlaybackUrlResponse>(
    `/api/clips/${encodeURIComponent(clipId)}/download-url`,
    {
      method: "POST",
      token,
      body: options ?? null,
    }
  );
}

// ============================================================================
// Thumbnail URL
// ============================================================================

/**
 * Get a short-lived thumbnail URL for a clip.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 * @returns Thumbnail URL
 */
export function getThumbnailUrl(clipId: string, token: string): Promise<DeliveryUrl> {
  return apiFetch<DeliveryUrl>(
    `/api/clips/${encodeURIComponent(clipId)}/thumbnail-url`,
    {
      method: "POST",
      token,
    }
  );
}

// ============================================================================
// Share Management
// ============================================================================

/**
 * Create or update a share link for a clip.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 * @param options - Share configuration
 * @returns Share response with public URL
 *
 * @example
 * ```ts
 * const share = await createShare("clip-123", authToken, {
 *   access_level: "view_playback",
 *   expires_in_hours: 24,
 * });
 * console.log("Share URL:", share.share_url);
 * ```
 */
export function createShare(
  clipId: string,
  token: string,
  options?: CreateShareRequest
): Promise<ShareResponse> {
  return apiFetch<ShareResponse>(`/api/clips/${encodeURIComponent(clipId)}/share`, {
    method: "POST",
    token,
    body: options ?? { access_level: "view_playback" },
  });
}

/**
 * Revoke a share link for a clip.
 *
 * After revocation, the share URL will return 410 Gone.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 */
export function revokeShare(clipId: string, token: string): Promise<void> {
  return apiFetch<void>(`/api/clips/${encodeURIComponent(clipId)}/share`, {
    method: "DELETE",
    token,
  });
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Check if a delivery URL is still valid (not expired).
 *
 * @param url - Delivery URL response
 * @param bufferSeconds - Buffer time before expiry (default: 60s)
 * @returns true if URL is still valid
 */
export function isUrlValid(
  url: DeliveryUrl | PlaybackUrlResponse,
  bufferSeconds = 60
): boolean {
  const expiresAt = new Date(url.expires_at);
  const now = new Date();
  const bufferMs = bufferSeconds * 1000;
  return expiresAt.getTime() > now.getTime() + bufferMs;
}

/**
 * Open a clip in a new tab for playback.
 *
 * NOTE: This function uses browser APIs (window) and must only be called
 * from client components. It will throw in SSR/edge contexts.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 */
export async function playInNewTab(clipId: string, token: string): Promise<void> {
  if (typeof window === "undefined") {
    throw new Error("playInNewTab can only be called in browser context");
  }
  const { url } = await getPlaybackUrl(clipId, token);
  window.open(url, "_blank");
}

/**
 * Trigger a clip download.
 *
 * NOTE: This function uses browser APIs (document) and must only be called
 * from client components. It will throw in SSR/edge contexts.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 * @param filename - Optional custom filename
 */
export async function downloadClip(
  clipId: string,
  token: string,
  filename?: string
): Promise<void> {
  if (typeof document === "undefined") {
    throw new Error("downloadClip can only be called in browser context");
  }
  const { url, clip } = await getDownloadUrl(clipId, token, { filename });
  const downloadName = filename ?? clip.filename;

  // Create a temporary anchor to trigger download
  const a = document.createElement("a");
  a.href = url;
  a.download = downloadName;
  a.style.display = "none";
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
}

/**
 * Copy a share URL to clipboard with toast feedback.
 *
 * Creates a share if needed and copies the URL. Shows sonner toast notifications
 * for success and failure states.
 *
 * NOTE: This function uses browser APIs (navigator.clipboard) and must only
 * be called from client components. It will silently fail in SSR/edge contexts.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 * @param options - Share configuration
 * @returns The share URL on success, or undefined on failure
 */
export async function copyShareUrl(
  clipId: string,
  token: string,
  options?: CreateShareRequest
): Promise<string | undefined> {
  // Guard against SSR / No Clipboard
  if (
    typeof window === "undefined" ||
    typeof navigator === "undefined" ||
    !navigator.clipboard
  ) {
    toast.error("Clipboard not available");
    return undefined;
  }

  const toastId = toast.loading("Generating share link...");

  try {
    // Create or get share
    const share = await createShare(clipId, token, options);

    // Copy to clipboard
    await navigator.clipboard.writeText(share.share_url);

    toast.success("Share link copied!", {
      id: toastId,
      description: "Anyone with this link can view the clip.",
    });

    return share.share_url;
  } catch (err) {
    console.error("[copyShareUrl] Failed:", err);
    toast.error("Failed to copy link", {
      id: toastId,
      description: err instanceof Error ? err.message : "Unknown error",
    });
    return undefined;
  }
}

/**
 * Copy a share URL to clipboard (throwing version).
 *
 * Like copyShareUrl but throws on error instead of showing toast.
 * Use this when you want to handle errors yourself.
 *
 * @param clipId - The clip ID
 * @param token - Auth token
 * @param options - Share configuration
 * @returns The share URL on success
 * @throws Error if clipboard is unavailable or access is denied
 */
export async function copyShareUrlOrThrow(
  clipId: string,
  token: string,
  options?: CreateShareRequest
): Promise<string> {
  if (typeof navigator === "undefined" || !navigator.clipboard) {
    throw new Error("Clipboard API is not available in this context");
  }

  const share = await createShare(clipId, token, options);

  try {
    await navigator.clipboard.writeText(share.share_url);
  } catch (err) {
    throw new Error(
      `Failed to copy to clipboard: ${err instanceof Error ? err.message : "Permission denied"}`
    );
  }

  return share.share_url;
}

File: backend/crates/vclip-api/src/handlers/clip_delivery.rs
//! Clip delivery handlers.
//!
//! Secure endpoints for clip playback, download, and sharing.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::Response;
use axum::Json;
use chrono::{Duration as ChronoDuration, Utc};
use serde::{Deserialize, Serialize};
use tracing::{info, warn};

use vclip_firestore::{ClipRepository, ShareRepository};
use vclip_models::{
    ClipStatus, CreateShareRequest, ShareConfig, ShareResponse,
    is_valid_share_slug, MAX_SHARE_EXPIRY_HOURS,
};
use vclip_storage::{DeliveryConfig, DeliveryUrl, DeliveryUrlGenerator};

use crate::auth::AuthUser;
use crate::error::{ApiError, ApiResult};
use crate::security::is_valid_clip_name;
use crate::state::AppState;

// ============================================================================
// Response Types
// ============================================================================

/// Response for playback/download URL requests.
#[derive(Debug, Serialize)]
pub struct PlaybackUrlResponse {
    /// The delivery URL.
    pub url: String,
    /// When this URL expires (ISO 8601).
    pub expires_at: String,
    /// Expiry in seconds from now.
    pub expires_in_secs: u64,
    /// Content type.
    pub content_type: String,
    /// Clip metadata.
    pub clip: ClipSummary,
}

/// Summary of clip metadata for delivery responses.
#[derive(Debug, Serialize)]
pub struct ClipSummary {
    pub clip_id: String,
    pub filename: String,
    pub title: String,
    pub duration_seconds: f64,
    pub file_size_bytes: u64,
}

/// Request body for download URL (optional filename override).
#[derive(Debug, Deserialize)]
pub struct DownloadUrlRequest {
    /// Custom filename for the download.
    #[serde(default)]
    pub filename: Option<String>,
}

// ============================================================================
// Playback URL Handler
// ============================================================================

/// Generate a short-lived playback URL for a clip.
///
/// POST /api/clips/{clip_id}/play-url
///
/// Response:
/// ```json
/// {
///   "url": "https://...",
///   "expires_at": "2024-01-01T12:15:00Z",
///   "expires_in_secs": 900,
///   "content_type": "video/mp4",
///   "clip": { ... }
/// }
/// ```
pub async fn get_playback_url(
    State(state): State<AppState>,
    Path(clip_id): Path<String>,
    user: AuthUser,
) -> ApiResult<Json<PlaybackUrlResponse>> {
    // Validate clip_id format
    if !is_valid_clip_name(&clip_id) {
        return Err(ApiError::bad_request("Invalid clip ID format"));
    }

    // Look up clip metadata
    let clip = find_clip_by_id(&state, &user.uid, &clip_id).await?;

    // Verify ownership
    if clip.user_id != user.uid {
        return Err(ApiError::forbidden("You don't own this clip"));
    }

    // Generate delivery URL
    let delivery_config = DeliveryConfig::from_env();
    let generator = DeliveryUrlGenerator::new((*state.storage).clone(), delivery_config);

    let delivery_url = generator
        .playback_url(&clip.r2_key, &clip.clip_id, &user.uid)
        .await
        .map_err(|e| {
            warn!(clip_id = %clip_id, error = %e, "Failed to generate playback URL");
            ApiError::internal("Failed to generate playback URL")
        })?;

    info!(clip_id = %clip_id, user_id = %user.uid, "Generated playback URL");

    Ok(Json(PlaybackUrlResponse {
        url: delivery_url.url,
        expires_at: delivery_url.expires_at,
        expires_in_secs: delivery_url.expires_in_secs,
        content_type: delivery_url.content_type,
        clip: ClipSummary {
            clip_id: clip.clip_id,
            filename: clip.filename,
            title: clip.scene_title,
            duration_seconds: clip.duration_seconds,
            file_size_bytes: clip.file_size_bytes,
        },
    }))
}

// ============================================================================
// Download URL Handler
// ============================================================================

/// Generate a short-lived download URL for a clip.
///
/// POST /api/clips/{clip_id}/download-url
///
/// Request body (optional):
/// ```json
/// { "filename": "my-custom-filename.mp4" }
/// ```
///
/// Response: Same as playback URL.
pub async fn get_download_url(
    State(state): State<AppState>,
    Path(clip_id): Path<String>,
    user: AuthUser,
    Json(body): Json<Option<DownloadUrlRequest>>,
) -> ApiResult<Json<PlaybackUrlResponse>> {
    // Validate clip_id format
    if !is_valid_clip_name(&clip_id) {
        return Err(ApiError::bad_request("Invalid clip ID format"));
    }

    // Look up clip metadata
    let clip = find_clip_by_id(&state, &user.uid, &clip_id).await?;

    // Verify ownership
    if clip.user_id != user.uid {
        return Err(ApiError::forbidden("You don't own this clip"));
    }

    // Determine filename for Content-Disposition
    let filename = body
        .as_ref()
        .and_then(|b| b.filename.as_deref())
        .map(sanitize_download_filename)
        .unwrap_or_else(|| clip.filename.clone());

    // Generate delivery URL
    let delivery_config = DeliveryConfig::from_env();
    let generator = DeliveryUrlGenerator::new((*state.storage).clone(), delivery_config);

    let delivery_url = generator
        .download_url(&clip.r2_key, &clip.clip_id, &user.uid, Some(&filename))
        .await
        .map_err(|e| {
            warn!(clip_id = %clip_id, error = %e, "Failed to generate download URL");
            ApiError::internal("Failed to generate download URL")
        })?;

    info!(clip_id = %clip_id, user_id = %user.uid, "Generated download URL");

    Ok(Json(PlaybackUrlResponse {
        url: delivery_url.url,
        expires_at: delivery_url.expires_at,
        expires_in_secs: delivery_url.expires_in_secs,
        content_type: delivery_url.content_type,
        clip: ClipSummary {
            clip_id: clip.clip_id,
            filename: clip.filename,
            title: clip.scene_title,
            duration_seconds: clip.duration_seconds,
            file_size_bytes: clip.file_size_bytes,
        },
    }))
}

// ============================================================================
// Thumbnail URL Handler
// ============================================================================

/// Generate a short-lived thumbnail URL for a clip.
///
/// POST /api/clips/{clip_id}/thumbnail-url
pub async fn get_thumbnail_url(
    State(state): State<AppState>,
    Path(clip_id): Path<String>,
    user: AuthUser,
) -> ApiResult<Json<DeliveryUrl>> {
    // Validate clip_id format
    if !is_valid_clip_name(&clip_id) {
        return Err(ApiError::bad_request("Invalid clip ID format"));
    }

    // Look up clip metadata
    let clip = find_clip_by_id(&state, &user.uid, &clip_id).await?;

    // Verify ownership
    if clip.user_id != user.uid {
        return Err(ApiError::forbidden("You don't own this clip"));
    }

    // Check if thumbnail exists
    let thumb_key = clip.thumbnail_r2_key.as_ref().ok_or_else(|| {
        ApiError::not_found("Clip has no thumbnail")
    })?;

    // Generate delivery URL
    let delivery_config = DeliveryConfig::from_env();
    let generator = DeliveryUrlGenerator::new((*state.storage).clone(), delivery_config);

    let delivery_url = generator
        .thumbnail_url(thumb_key, &clip.clip_id, &user.uid)
        .await
        .map_err(|e| {
            warn!(clip_id = %clip_id, error = %e, "Failed to generate thumbnail URL");
            ApiError::internal("Failed to generate thumbnail URL")
        })?;

    Ok(Json(delivery_url))
}

// ============================================================================
// Share Handlers
// ============================================================================

/// Create or update a share link for a clip.
///
/// POST /api/clips/{clip_id}/share
///
/// Request body:
/// ```json
/// {
///   "access_level": "view_playback",
///   "expires_in_hours": 24,
///   "watermark_enabled": false
/// }
/// ```
///
/// Response:
/// ```json
/// {
///   "share_url": "https://viralclipai.io/c/abc123xyz",
///   "share_slug": "abc123xyz",
///   "access_level": "view_playback",
///   "expires_at": "2024-01-02T12:00:00Z",
///   "watermark_enabled": false,
///   "created_at": "2024-01-01T12:00:00Z"
/// }
/// ```
pub async fn create_share(
    State(state): State<AppState>,
    Path(clip_id): Path<String>,
    user: AuthUser,
    Json(body): Json<CreateShareRequest>,
) -> ApiResult<Json<ShareResponse>> {
    // Validate clip_id format
    if !is_valid_clip_name(&clip_id) {
        return Err(ApiError::bad_request("Invalid clip ID format"));
    }

    // Look up clip metadata
    let clip = find_clip_by_id(&state, &user.uid, &clip_id).await?;

    // Verify ownership
    if clip.user_id != user.uid {
        return Err(ApiError::forbidden("Only the owner can share this clip"));
    }

    // Validate expiry if specified (max 30 days)
    if let Some(hours) = body.expires_in_hours {
        if hours > MAX_SHARE_EXPIRY_HOURS {
            return Err(ApiError::bad_request(format!(
                "Expiry cannot exceed {} hours ({} days)",
                MAX_SHARE_EXPIRY_HOURS,
                MAX_SHARE_EXPIRY_HOURS / 24
            )));
        }
    }

    // Create share config
    let mut share_config = ShareConfig::new(
        &clip.clip_id,
        &user.uid,
        clip.video_id.as_str(),
        body.access_level,
    );

    // Set expiry if specified
    if let Some(hours) = body.expires_in_hours {
        let expires_at = Utc::now() + ChronoDuration::hours(hours as i64);
        share_config = share_config.with_expiry(expires_at);
    }

    // Set watermark if enabled
    if body.watermark_enabled {
        share_config = share_config.with_watermark();
    }

    // Persist share config to Firestore (dual-document pattern)
    let share_repo = ShareRepository::new((*state.firestore).clone());
    share_repo.create_share(&share_config).await.map_err(|e| {
        warn!(clip_id = %clip_id, error = %e, "Failed to persist share config");
        ApiError::internal("Failed to create share link")
    })?;

    let base_url = std::env::var("PUBLIC_APP_URL")
        .unwrap_or_else(|_| "https://viralclipai.io".to_string());

    let response = ShareResponse::from_config(&share_config, &base_url);

    info!(
        clip_id = %clip_id,
        user_id = %user.uid,
        share_slug = %share_config.share_slug,
        "Created share link"
    );

    Ok(Json(response))
}

/// Revoke a share link for a clip.
///
/// DELETE /api/clips/{clip_id}/share
///
/// Response: 204 No Content on success.
pub async fn revoke_share(
    State(state): State<AppState>,
    Path(clip_id): Path<String>,
    user: AuthUser,
) -> ApiResult<StatusCode> {
    // Validate clip_id format
    if !is_valid_clip_name(&clip_id) {
        return Err(ApiError::bad_request("Invalid clip ID format"));
    }

    // Look up clip metadata
    let clip = find_clip_by_id(&state, &user.uid, &clip_id).await?;

    // Verify ownership
    if clip.user_id != user.uid {
        return Err(ApiError::forbidden("Only the owner can revoke sharing"));
    }

    // Get the share config to find the slug
    let share_repo = ShareRepository::new((*state.firestore).clone());
    let share_config = share_repo
        .get_config(&user.uid, clip.video_id.as_str(), &clip_id)
        .await
        .map_err(|e| {
            warn!(clip_id = %clip_id, error = %e, "Failed to look up share config");
            ApiError::internal("Database error")
        })?;

    // If no share config exists, return success (idempotent)
    let config = match share_config {
        Some(c) => c,
        None => {
            info!(clip_id = %clip_id, user_id = %user.uid, "No active share to revoke");
            return Ok(StatusCode::NO_CONTENT);
        }
    };

    // Disable the share (updates config and deletes slug index)
    share_repo
        .disable_share(&user.uid, clip.video_id.as_str(), &clip_id, &config.share_slug)
        .await
        .map_err(|e| {
            warn!(clip_id = %clip_id, error = %e, "Failed to revoke share");
            ApiError::internal("Failed to revoke share link")
        })?;

    info!(
        clip_id = %clip_id,
        user_id = %user.uid,
        share_slug = %config.share_slug,
        "Revoked share link"
    );

    Ok(StatusCode::NO_CONTENT)
}

// ============================================================================
// Public Share Resolution (for /c/{share_slug} redirect)
// ============================================================================

/// Resolve a share slug to a playback redirect.
///
/// GET /c/{share_slug}
///
/// Returns: 302 redirect to a short-lived presigned URL, or error.
pub async fn resolve_share(
    State(state): State<AppState>,
    Path(share_slug): Path<String>,
) -> Result<Response, ApiError> {
    use axum::http::header;
    use axum::response::IntoResponse;
    use std::time::Duration;

    // Validate share_slug format
    if !is_valid_share_slug(&share_slug) {
        return Err(ApiError::bad_request("Invalid share link"));
    }

    // Look up share by slug
    let share_repo = ShareRepository::new((*state.firestore).clone());
    let slug_info = share_repo.get_by_slug(&share_slug).await.map_err(|e| {
        warn!(share_slug = %share_slug, error = %e, "Failed to look up share slug");
        ApiError::internal("Database error")
    })?;

    // Check if share exists
    let slug_info = match slug_info {
        Some(info) => info,
        None => {
            info!(share_slug = %share_slug, "Share slug not found");
            return Err(ApiError::NotFound("Share link not found".to_string()));
        }
    };

    // Check if share is disabled
    if slug_info.disabled_at.is_some() {
        info!(share_slug = %share_slug, "Share has been revoked");
        return Err(ApiError::Gone("Share link has been revoked".to_string()));
    }

    // Check if share is expired
    if let Some(expires_at) = slug_info.expires_at {
        if Utc::now() > expires_at {
            info!(share_slug = %share_slug, "Share has expired");
            return Err(ApiError::Gone("Share link has expired".to_string()));
        }
    }

    // Look up clip metadata to get R2 key
    let clip = find_clip_by_owner_context(
        &state,
        &slug_info.user_id,
        &slug_info.video_id,
        &slug_info.clip_id,
    )
    .await?;

    // Generate short-lived delivery URL (1 hour)
    let delivery_config = DeliveryConfig::from_env();
    let generator = DeliveryUrlGenerator::new((*state.storage).clone(), delivery_config.clone());

    // Use Worker URL if configured, otherwise presigned URL
    let delivery_url = if delivery_config.should_use_worker() {
        generator
            .generate_worker_url_with_key(
                &clip.clip_id,
                &slug_info.user_id,
                Some(&clip.r2_key),
                vclip_storage::DeliveryScope::Playback,
                Duration::from_secs(3600), // 1 hour
            )
            .map_err(|e| {
                warn!(share_slug = %share_slug, error = %e, "Failed to generate worker URL");
                ApiError::internal("Failed to generate playback URL")
            })?
    } else {
        generator
            .playback_url(&clip.r2_key, &clip.clip_id, &slug_info.user_id)
            .await
            .map_err(|e| {
                warn!(share_slug = %share_slug, error = %e, "Failed to generate playback URL");
                ApiError::internal("Failed to generate playback URL")
            })?
    };

    info!(
        share_slug = %share_slug,
        clip_id = %clip.clip_id,
        user_id = %slug_info.user_id,
        "Resolved share link"
    );

    // Build 302 redirect response with cache control
    let response = (
        StatusCode::FOUND,
        [
            (header::LOCATION, delivery_url.url),
            // Prevent caching so revocation takes effect quickly
            (header::CACHE_CONTROL, "private, max-age=60".to_string()),
        ],
    )
        .into_response();

    Ok(response)
}

/// Find a clip by owner context (user_id, video_id, clip_id).
async fn find_clip_by_owner_context(
    state: &AppState,
    user_id: &str,
    video_id: &str,
    clip_id: &str,
) -> ApiResult<vclip_models::ClipMetadata> {
    let video_id_obj = vclip_models::VideoId::from_string(video_id.to_string());
    let clip_repo = ClipRepository::new((*state.firestore).clone(), user_id, video_id_obj);

    let clips = clip_repo.list(Some(ClipStatus::Completed)).await.map_err(|e| {
        warn!(video_id = %video_id, error = %e, "Failed to list clips");
        ApiError::internal("Database error")
    })?;

    clips
        .into_iter()
        .find(|c| c.clip_id == clip_id)
        .ok_or_else(|| ApiError::not_found("Clip not found"))
}

// ============================================================================
// Helper Functions
// ============================================================================

/// Find a clip by ID, searching across all user's videos.
///
/// This is a temporary implementation that lists all videos and clips.
/// In a production system, we'd have a direct clip lookup by ID.
async fn find_clip_by_id(
    state: &AppState,
    user_id: &str,
    clip_id: &str,
) -> ApiResult<vclip_models::ClipMetadata> {
    // List user's videos
    let video_repo = vclip_firestore::VideoRepository::new(
        (*state.firestore).clone(),
        user_id,
    );

    let videos = video_repo.list(Some(100)).await.map_err(|e| {
        warn!(user_id = %user_id, error = %e, "Failed to list videos");
        ApiError::internal("Database error")
    })?;

    // Search clips in each video
    for video in videos {
        let clip_repo = ClipRepository::new(
            (*state.firestore).clone(),
            user_id,
            video.video_id.clone(),
        );

        let clips = clip_repo.list(Some(ClipStatus::Completed)).await.map_err(|e| {
            warn!(video_id = %video.video_id, error = %e, "Failed to list clips");
            ApiError::internal("Database error")
        })?;

        if let Some(clip) = clips.into_iter().find(|c| c.clip_id == clip_id) {
            return Ok(clip);
        }
    }

    Err(ApiError::not_found("Clip not found"))
}

/// Sanitize a download filename.
fn sanitize_download_filename(name: &str) -> String {
    let safe_joined: String = name
        .split(|c| c == '/' || c == '\\')
        .filter(|s| !s.is_empty() && *s != "." && *s != "..")
        .collect::<Vec<_>>()
        .join("");

    let sanitized: String = safe_joined
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || *c == '-' || *c == '_' || *c == '.')
        .take(100)
        .collect();

    if sanitized.is_empty() {
        "clip.mp4".to_string()
    } else if !sanitized.ends_with(".mp4") {
        format!("{}.mp4", sanitized)
    } else {
        sanitized
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitize_download_filename() {
        assert_eq!(sanitize_download_filename("my-video.mp4"), "my-video.mp4");
        assert_eq!(sanitize_download_filename("my video"), "myvideo.mp4");
        assert_eq!(sanitize_download_filename(""), "clip.mp4");
        assert_eq!(sanitize_download_filename("../../../etc/passwd"), "etcpasswd.mp4");
    }
}
